File: ung_dung_diem_danh/lib/models/diem_danh_request.dart
import 'package:equatable/equatable.dart';

class DiemDanhVaoRequest extends Equatable {
  final String maNhanVien;
  final String phuongThuc;
  final double? viDo;
  final double? kinhDo;
  final String? ghiChu;

  const DiemDanhVaoRequest({
    required this.maNhanVien,
    this.phuongThuc = 'SinhTracHoc',
    this.viDo,
    this.kinhDo,
    this.ghiChu,
  });

  Map<String, dynamic> toJson() {
    return {
      'maNhanVien': maNhanVien.toString(),
      'phuongThuc': phuongThuc,
      'viDo': viDo,
      'kinhDo': kinhDo,
      'ghiChu': ghiChu,
    };
  }

  @override
  List<Object?> get props => [maNhanVien, phuongThuc, viDo, kinhDo, ghiChu];
}

class DiemDanhRaRequest extends Equatable {
  final String maNhanVien;
  final String phuongThuc;
  final double? viDo;
  final double? kinhDo;
  final String? ghiChu;

  const DiemDanhRaRequest({
    required this.maNhanVien,
    this.phuongThuc = 'SinhTracHoc',
    this.viDo,
    this.kinhDo,
    this.ghiChu,
  });

  Map<String, dynamic> toJson() {
    return {
      'maNhanVien': maNhanVien.toString(),
      'phuongThuc': phuongThuc,
      'viDo': viDo,
      'kinhDo': kinhDo,
      'ghiChu': ghiChu,
    };
  }

  @override
  List<Object?> get props => [maNhanVien, phuongThuc, viDo, kinhDo, ghiChu];
}

class DiemDanhThuCongRequest extends Equatable {
  final String maNhanVien;
  final DateTime gioVao;
  final DateTime? gioRa;
  final String? ghiChu;

  const DiemDanhThuCongRequest({
    required this.maNhanVien,
    required this.gioVao,
    this.gioRa,
    this.ghiChu,
  });

  Map<String, dynamic> toJson() {
    return {
      'maNhanVien': maNhanVien.toString(),
      'gioVao': gioVao.toIso8601String(),
      'gioRa': gioRa?.toIso8601String(),
      'ghiChu': ghiChu,
    };
  }

  @override
  List<Object?> get props => [maNhanVien, gioVao, gioRa, ghiChu];
}

File: UngDungDiemDanhNhanVien/DTOs/DiemDanhRequest.cs
using System.ComponentModel.DataAnnotations;

namespace UngDungDiemDanhNhanVien.DTOs
{
    public class DiemDanhVaoRequest
    {
        [Required(ErrorMessage = "Mã nhân viên là bắt buộc")]
        public string MaNhanVien { get; set; } = string.Empty;

        [StringLength(50, ErrorMessage = "Phương thức không được vượt quá 50 ký tự")]
        public string PhuongThuc { get; set; } = "SinhTracHoc"; // SinhTracHoc, KhuonMat, ThuCong

        [Range(-90, 90, ErrorMessage = "Vĩ độ không hợp lệ")]
        public decimal? ViDo { get; set; }

        [Range(-180, 180, ErrorMessage = "Kinh độ không hợp lệ")]
        public decimal? KinhDo { get; set; }

        [StringLength(500, ErrorMessage = "Ghi chú không được vượt quá 500 ký tự")]
        public string? GhiChu { get; set; }
    }

    public class DiemDanhRaRequest
    {
        [Required(ErrorMessage = "Mã nhân viên là bắt buộc")]
        public string MaNhanVien { get; set; } = string.Empty;

        [StringLength(50, ErrorMessage = "Phương thức không được vượt quá 50 ký tự")]
        public string PhuongThuc { get; set; } = "SinhTracHoc"; // SinhTracHoc, KhuonMat, ThuCong

        [Range(-90, 90, ErrorMessage = "Vĩ độ không hợp lệ")]
        public decimal? ViDo { get; set; }

        [Range(-180, 180, ErrorMessage = "Kinh độ không hợp lệ")]
        public decimal? KinhDo { get; set; }

        [StringLength(500, ErrorMessage = "Ghi chú không được vượt quá 500 ký tự")]
        public string? GhiChu { get; set; }
    }

    public class DiemDanhThuCongRequest
    {
        [Required(ErrorMessage = "ID nhân viên là bắt buộc")]
        public int NhanVienId { get; set; }

        [Required(ErrorMessage = "Giờ vào là bắt buộc")]
        public DateTime GioVao { get; set; }

        public DateTime? GioRa { get; set; }

        [Range(-90, 90, ErrorMessage = "Vĩ độ không hợp lệ")]
        public decimal? ViDo { get; set; }

        [Range(-180, 180, ErrorMessage = "Kinh độ không hợp lệ")]
        public decimal? KinhDo { get; set; }

        [StringLength(500, ErrorMessage = "Ghi chú không được vượt quá 500 ký tự")]
        public string? GhiChu { get; set; }
    }

    public class DiemDanhNfcRequest
    {
        [Required(ErrorMessage = "Mã thẻ NFC là bắt buộc")]
        public string MaTheNfc { get; set; } = string.Empty;

        [Range(-90, 90, ErrorMessage = "Vĩ độ không hợp lệ")]
        public decimal? ViDo { get; set; }

        [Range(-180, 180, ErrorMessage = "Kinh độ không hợp lệ")]
        public decimal? KinhDo { get; set; }

        [StringLength(500, ErrorMessage = "Ghi chú không được vượt quá 500 ký tự")]
        public string? GhiChu { get; set; }
    }
}